### ISSUE TITLE
IntegrationTestCase can't test a route that uses RedirectRoute.

### ISSUE THREAD

**Issue by beporter** [cakephp/cakephp#8015]
tl;dr: If you have a `$route->redirect('/something', [...]);` in your `config/routes.php` and try `$this->get('/something');` in a unit test, PHPUnit will exit with status code zero (cleanly) at that point, which could be a problem for automated tests since it will look like "success".

Steps to reproduce:

1. Set up a new Cake app:

	```shell
	$ composer create-project cakephp/app redir_test/
	Installing cakephp/app (3.1.2)
	  - Installing cakephp/app (3.1.2)
		Downloading: 100%         

	Created project in /redir_test
	Loading composer repositories with package information
	Installing dependencies (including require-dev)

	// *snip*

	$ cd redir_test/
	```

2. Create a controller at `/redir_test/src/Controller/TestsController.php`:

	```php
	<?php
	namespace App\Controller;
	use App\Controller\AppController;
	class TestsController extends AppController {
		public function test() {
			$this->set('foo', 'bar');
		}
	}
	```

3. Create a test at `/redir_test/tests/TestCase/Controller/TestsControllerTest.php`:

	```php
	<?php
	namespace App\Test\TestCase\Controller;
	use App\Controller\TestsController;
	use Cake\TestSuite\IntegrationTestCase;

	class TestsControllerTest extends IntegrationTestCase {
		public function testTest() {
			$this->get('/');
			$this->assertResponseOk(); // <-- We will never see this.
		}
	}
	```

4. Add a redirect in `config/routes.php`:

	```diff
	 Router::scope('/', function ($routes) {
		 /**
		  * Here, we are connecting '/' (base path) to a controller called 'Pages',
		  * its action called 'display', and we pass a param to select the view file
		  * to use (in this case, src/Template/Pages/home.ctp)...
		  */
	+    $routes->redirect('/', ['controller' => 'Tests', 'action' => 'test']);
		 $routes->connect('/', ['controller' => 'Pages', 'action' => 'display', 'home']);
	```

5. Try running the new test:

	```shell
	$ vendor/bin/phpunit tests/TestCase/Controller/TestsControllerTest.php
	PHPUnit 5.1.4 by Sebastian Bergmann and contributors.


	// (^^^ There should be 1 dot here ^^^)

	$
	```

---

Tracing the code bring you to [Routing/Route/RedirectRoute.php L99](https://github.com/cakephp/cakephp/blob/3.1.7/src/Routing/Route/RedirectRoute.php#L99):
```php
        $this->response->statusCode($status);
        $this->response->send();
        $this->response->stop();
```

And of course, [Network/Response.php L1612-1615](https://github.com/cakephp/cakephp/blob/3.1.7/src/Network/Response.php#L1612,L1615) calls `exit(0)`:
```php
    public function stop($status = 0)
    {
        exit($status);
    }
```

The `IntegrationTestCase` doesn't mock this at all, so PHPUnit just silently stops whenever the redirected URL is hit. The obvious workaround is to be careful never to touch a redirect route from your tests, but ideally this should be more durable. I've been trying to work out if there's a way to spy on this particular event and capture the Response object instead so a unit test could check for the `redirect()`.

There is a `src/TestSuite/Stub/Response` class, but it's not used during the `RedirectRoute::parse()` so it's no help in intercepting the exit().

Here's the stack trace, for reference:
```
Cake\Network\Response::stop() - CORE/src/Network/Response.php, line 1614
Cake\Routing\Route\RedirectRoute::parse() - CORE/src/Routing/Route/RedirectRoute.php, line 100
Cake\Routing\RouteCollection::parse() - CORE/src/Routing/RouteCollection.php, line 125
Cake\Routing\Router::parse() - CORE/src/Routing/Router.php, line 339
Cake\Routing\Filter\RoutingFilter::beforeDispatch() - CORE/src/Routing/Filter/RoutingFilter.php, line 54
Cake\Routing\DispatcherFilter::handle() - CORE/src/Routing/DispatcherFilter.php, line 146
Cake\Event\EventManager::_callListener() - CORE/src/Event/EventManager.php, line 391
Cake\Event\EventManager::dispatch() - CORE/src/Event/EventManager.php, line 356
Cake\Routing\Dispatcher::dispatchEvent() - CORE/src/Event/EventDispatcherTrait.php, line 78
Cake\Routing\Dispatcher::dispatch() - CORE/src/Routing/Dispatcher.php, line 63
Cake\TestSuite\IntegrationTestCase::_sendRequest() - CORE/src/TestSuite/IntegrationTestCase.php, line 322
Cake\TestSuite\IntegrationTestCase::get() - CORE/src/TestSuite/IntegrationTestCase.php, line 233
App\Test\TestCase\Controller\TestsControllerTest::testTest() - ROOT/tests/TestCase/Controller/TestsControllerTest.php, line 11
```

Suffice to say, the approach for fixing this (if there is one) isn't obvious to me and I at least want to get this documented.

**Comment by markstory** [comment]
The `RedirectRoute` creates its own Response object which is the source of trouble here. I can't think of any completely backwards compatible way of solving this specific issue. However, if `RedirectRoute` were to start raising a `RedirectException` to signal redirects, we could handle this situation with minimal breaking changes.

**Comment by lorenzo** [comment]
I have worked around this issue in my projects by setting a unique param in the request object having the full URL I want to redirect to.

Later, another dispatcher filter reads the param and stops the dispatch I got oceans after setting the redirect header. 

We could use sometime similar to that, where the routing dispatcher filter reads the special param and returns the response with the redirect header set.

**Comment by markstory** [comment]
I spent a few minutes on this tonight, and have [this diff as a proof of concept](https://github.com/cakephp/cakephp/compare/3.2...issue-8015). I think it solves the problem outlined here, and cleans up RedirectRoute a bit.

**Comment by jeremyharris** [comment]
Nice elegant fix @markstory.

**Comment by markstory** [comment]
Closing as a pull request is open now.
